src/formatter.rs

```diff
@@ -83,13 +117,119 @@ where
                 command = title;
                 debug!("Detected command name: {}", &command);
             }
-            LineType::Description(text) => println!("  {}", config.style.description.paint(text)),
-            LineType::ExampleText(text) => println!("  {}", config.style.example_text.paint(text)),
+            LineType::Description(text) => yield_snippet(HighlightingSnippet::Description(&text)),
+            LineType::ExampleText(text) => yield_snippet(HighlightingSnippet::Text(&text)),
             LineType::ExampleCode(text) => {
-                println!("      {}", &format_code(&command, &text, &config))
+                yield_snippet(HighlightingSnippet::NormalCode("      "));
+                highlight_code(&command, &text, &mut yield_snippet);
+                yield_snippet(HighlightingSnippet::Linebreak);
             }
+
             LineType::Other(text) => debug!("Unknown line type: {:?}", text),
         }
     }
-    println!();
+    yield_snippet(HighlightingSnippet::Linebreak);
+}
+
+fn print_snippet<'a>(snip: HighlightingSnippet<'a>, style: &StyleConfig) {
```

Comment:
In object-oriented fashion this function could be added as a method to `HighlightingSnippet`.

Reply:
Yeah, that thought crossed my mind too, but in the end I decided that I would like `HighlighttingSnippet` to be just a "stupid" data type rather than an active player. I found it weird that it would "know" the `StyleConfig` type somehow and I didn't want to encode how to use the type in it's definition but have `print_lines` tie the knots between this whole module and the tealdeer cli.

Is there another (OO) intuition that I don't see here and fits this whole module better?

Reply:
To get this really clean either `print_lines` and `print_snippet` should be moved to `main.rs` so that the formatting module does not even need to know `StyleConfig` (and the `Tokenizer`) or `StyleConfig` should be defined in `formatter.rs` itself so it can be used freely inside this module (thus `print_snippet` can be a method of `HighlightingSnippet`). Otherwise, the formatting code will not really be separated from the rest.

From a testing perspective, dealing with `StyleConfig` is more work, however, so getting rid of it wherever possible might be preferred.

Reply:
@SimplyDanny I managed to extract the printing from `main.rs` and `formatter.rs` (for details see commit messages)